state/watcher.go

```diff
@@ -496,6 +497,154 @@ func watchApplicationRelations(backend modelBackend, applicationName string) Str
 	return newRelationLifeSuspendedWatcher(backend, members, filter, nil)
 }
 
+// WatchModelMachineStartTimes watches the non-container machines in the model
+// for changes to the Life or AgentStartTime fields and reports them as a batch
+// after the specified quiesceInterval time has passed without seeing any new
+// change events.
+func (st *State) WatchModelMachineStartTimes(quiesceInterval time.Duration) StringsWatcher {
+	return newModelMachineStartTimeWatcher(st, st.clock(), quiesceInterval)
+}
+
+type modelMachineStartTimeWatcher struct {
+	commonWatcher
+	outCh chan []string
+
+	clk             clock.Clock
+	quiesceInterval time.Duration
+	docHash         map[string]string
+}
+
+func newModelMachineStartTimeWatcher(backend modelBackend, clk clock.Clock, quiesceInterval time.Duration) StringsWatcher {
+	w := &modelMachineStartTimeWatcher{
+		commonWatcher:   newCommonWatcher(backend),
+		outCh:           make(chan []string),
+		clk:             clk,
+		quiesceInterval: quiesceInterval,
+		docHash:         make(map[string]string),
+	}
+
+	w.tomb.Go(func() error {
+		defer close(w.outCh)
+		return w.loop()
+	})
+	return w
+}
+
+// SetClock replaces the clock implementation used by the watcher. This method
+// is meant to be used by tests.
+func (w *modelMachineStartTimeWatcher) SetClock(clk clock.Clock) {
+	w.clk = clk
+}
+
+// Changes returns the event channel for the watcher.
+func (w *modelMachineStartTimeWatcher) Changes() <-chan []string {
+	return w.outCh
+}
+
+func (w *modelMachineStartTimeWatcher) loop() error {
+	docWatcher := newCollectionWatcher(w.backend, colWCfg{col: machinesC})
+	defer func() { _ = docWatcher.Stop() }()
+
+	var (
+		timer           clock.Timer
+		timerCh         <-chan time.Time
+		unprocessedDocs = make(set.Strings)
+		outCh           chan []string
+		changeSet       []string
+	)
+
+	defer func() {
+		if timer != nil {
+			_ = timer.Stop()
+		}
+	}()
+
+	for {
+		select {
+		case <-w.tomb.Dying():
+			return tomb.ErrDying
+		case <-w.watcher.Dead():
+			return stateWatcherDeadError(w.watcher.Err())
+		case changes := <-docWatcher.Changes():
+			for _, docID := range changes {
+				unprocessedDocs.Add(w.backend.docID(docID))
+			}
+
+			// Set up timer or reset it if already exists
+			if timer != nil {
+				timer.Reset(w.quiesceInterval)
```

Comment:
TBH I actually hadn't thought of broken agents that keep bouncing. I would assume that broken agents would be something that we caught during QA and  hopefully not try to debug on a live system...

Having said that, the current poll interval for entries in the instancepoller's short poll group is currently set to 3s. When the 2.7 instancepoller receives a change it will move the entries (if not already there) to the short poll group and **reset** their (individual) next poll intervals to 3s. So if the misbehaving worker takes less than 3s between restarts we will keep reseting the poll interval for it and not hit the cloud API (we will still access the API for non-crashing ones).

We could add extra logic to the process method which keeps track of a lastModifiedAt timestamp in the fieldDoc and use that for debouncing but I think it's probably an overkill for this application. Thoughts?

The suggestion to remove the timer reset code for each incoming change is indeed valid so I will go ahead and apply that.

Reply:
Well, sure ideally we would catch a bug like that! :) But we've seen weird problems in the past in production systems.

I think you're right that the extra logic is overkill. I'm less worried that we won't ever get to poll that one dodgy machine's address than that the bouncing machine means we don't refresh some other machine's address.